[ iOS ] imported/w3c/web-platform-tests/screen-orientation/fullscreen-interactions.html is a flaky failure

marcoscaceres opened the pull request:
#### 81ceb06a3201a978675437250f8b2c9e38b64c15
<pre>
[ iOS ] imported/w3c/web-platform-tests/screen-orientation/fullscreen-interactions.html is a flaky failure
<a href="https://bugs.webkit.org/show_bug.cgi?id=256671">https://bugs.webkit.org/show_bug.cgi?id=256671</a>
rdar://109537056

Reviewed by Chris Dumez.

Refactored the test so to make it less flaky. Tested it a few thousand times locally.

* LayoutTests/imported/w3c/web-platform-tests/screen-orientation/fullscreen-interactions-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/screen-orientation/fullscreen-interactions.html:
* LayoutTests/imported/w3c/web-platform-tests/screen-orientation/resources/orientation-utils.js:
(export.async attachIframe):
* LayoutTests/platform/ios-wk2/TestExpectations:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/screen-orientation/fullscreen-interactions-expected.txt:
* LayoutTests/platform/mac-wk2/TestExpectations:

Canonical link: <a href="https://commits.webkit.org/264830@main">https://commits.webkit.org/264830@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2e922cfb9cf7f043d5db9517578c0285d2f735bd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8732 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/9020 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/9237 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/10385 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8720 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/11007 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8990 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11587 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8878 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9885 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7772 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/10549 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7182 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7981 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/15497 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/8292 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/8129 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/11475 "8 api tests failed or timed out") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8612 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/7017 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7873 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/7889 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2125 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12086 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8359 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

marcoscaceres commented on LayoutTests/imported/w3c/web-platform-tests/screen-orientation/fullscreen-interactions.html:
```diff
@@ -5,29 +5,85 @@
+    await fsExitPromise;
+  }, "Exiting fullscreen from an iframe causes a pending lock to be aborted");
+
+  // promise_test(async (t) => {
```
I commented this test out on purpose. This test would work if `orientationManager->unlock()` would work. That should cause the orientation to change and the event to fire. 

Right now, this test times out as the event never fires.

cdumez replied:
Doesn't seem right to keep a test commented out. Will the test be commented out upstream as well? We can't diverge from upstream.

marcoscaceres replied:
Nah, was just temporary as I was trying to figure out why calling `orientationManager->unlock()` doesn't work.

cdumez reviewed (commented):
Is this a WPT resync? All test changes will get overwritten unless we upstream these changes.

cdumez commented on LayoutTests/imported/w3c/web-platform-tests/screen-orientation/active-lock-expected.txt:
```diff
@@ -1,6 +1,7 @@
+CONSOLE MESSAGE: Unhandled Promise Rejection: NotSupportedError: Screen orientation locking is not supported
```
Why is screen orientation locking not supported? It should be enabled in the tests.

marcoscaceres replied:
It's only on iOS, right?

cdumez replied:
But this patch doesn't update test expectations on iOS? Why? And why is the iOS EWS green if the test output didn't get updated?

cdumez commented on LayoutTests/imported/w3c/web-platform-tests/screen-orientation/fullscreen-interactions-expected.txt:
```diff
@@ -1,5 +1,6 @@
-
-FAIL Fully unlocking the screen orientation causes a pending lock to be aborted promise_rejects_dom: function "function () { throw e }" threw object "NotSupportedError: Screen orientation locking is not supported" that is not a DOMException AbortError: property "code" is equal to 9, expected 20
-FAIL Fully unlocking the screen orientation causes a pending lock in a nested browsing context to be aborted promise_rejects_dom: function "function () { throw e }" threw object "SecurityError: Locking the screen orientation is only allowed when in fullscreen" that is not a DOMException AbortError: property "code" is equal to 18, expected 20
+FAIL fullscreen and orientation support promise_test: Unhandled rejection with value: object "NotSupportedError: Screen orientation locking is not supported"
```
ditto.

marcoscaceres replied:
As above... lock() only on iOS, right?

cdumez commented on Source/WebCore/page/ScreenOrientation.cpp:
```diff
@@ -118,7 +118,7 @@ void ScreenOrientation::lock(LockType lockType, Ref<DeferredPromise>&& promise)
     if (document->settings().fullscreenRequirementForScreenOrientationLockingEnabled()) {
 #if ENABLE(FULLSCREEN_API)
-        if (!document->fullscreenManager().isFullscreen()) {
+        if (!document->fullscreenManager().fullscreenElement()) {
```
Old code seemed clearer?

marcoscaceres replied:
This method more correct according to @nt1m. We should really rename `isFullscreen()` to something else to avoid confusion.

cdumez replied:
It is not more correct. They are identical but I think the old name is clearer.

nt1m replied:
@cdumez They are not identical when content inside an iframe is fullscreened, isFullscreen() will return false on the ancestor document, while fullscreenElement() will return the iframe element being implicitly fullscreened.

isFullscreen() and currentFullscreenElement() are identical however. It’s been this way for quite a long time, but I think it can probably be cleaned up.

cdumez replied:
Oh I see, `fullscreenElement()` doesn't actually return `m_fullscreenElement`. A little surprising but OK. Seems this code could use some refactoring / renaming for clarity.

cdumez replied:
Either way, it is a good idea to make this change separately, with a test.

marcoscaceres replied:
will do.

cdumez commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -378,6 +380,18 @@ void FullscreenManager::exitFullscreen(RefPtr<DeferredPromise>&& promise)
+        // https://w3c.github.io/screen-orientation/#dfn-fully-unlock-the-screen-orientation-steps
+        if (auto orientationManager = page->screenOrientationManager()) {
+            if (auto orientationPromise = orientationManager->takeLockPromise()) {
+                m_document.eventLoop().queueTask(TaskSource::UserInteraction, [orientationPromise = WTFMove(orientationPromise)]() {
```
`()` is not necessary.

cdumez commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -378,6 +380,18 @@ void FullscreenManager::exitFullscreen(RefPtr<DeferredPromise>&& promise)
 
+        // Run the "fully unlock the screen orientation" steps with doc.
+        // https://w3c.github.io/screen-orientation/#dfn-fully-unlock-the-screen-orientation-steps
+        if (auto orientationManager = page->screenOrientationManager()) {
```
We should split code changes out from test changes. The tests are changing so much in this PR, it is difficult to see what the progression from this change is.

marcoscaceres replied:
Yeah, my apologies for that. This was more for discussion. My intent was to split this into several PRs. 

I'll do that now.

cdumez replied:
Oh, I see. Bri really wanted to land this and asked me to review. I didn't realize this wasn't ready.

cdumez commented on LayoutTests/platform/mac-wk2/TestExpectations:
```diff
@@ -1792,6 +1792,4 @@ webkit.org/b/256189 [ Debug ] ipc/wait-for-video-output-will-change.html [ Pass
 
 webkit.org/b/256217 [ Debug ] imported/w3c/web-platform-tests/background-fetch/port-blocking.https.window.html [ Pass Crash ]
 
-webkit.org/b/256613 imported/w3c/web-platform-tests/screen-orientation/fullscreen-interactions.html [ DumpJSConsoleLogInStdErr ]
```
Can we drop this line too?
LayoutTests/platform/ios-wk2/TestExpectations:webkit.org/b/256613 imported/w3c/web-platform-tests/screen-orientation/fullscreen-interactions.html [ DumpJSConsoleLogInStdErr ]

marcoscaceres committed:
[ iOS ] imported/w3c/web-platform-tests/screen-orientation/fullscreen-interactions.html is a flaky failure
https://bugs.webkit.org/show_bug.cgi?id=256671
rdar://109537056

Reviewed by Chris Dumez.

Refactored the test so to make it less flaky. Tested it a few thousand times locally.

* LayoutTests/imported/w3c/web-platform-tests/screen-orientation/fullscreen-interactions-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/screen-orientation/fullscreen-interactions.html:
* LayoutTests/imported/w3c/web-platform-tests/screen-orientation/resources/orientation-utils.js:
(export.async attachIframe):
* LayoutTests/platform/ios-wk2/TestExpectations:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/screen-orientation/fullscreen-interactions-expected.txt:
* LayoutTests/platform/mac-wk2/TestExpectations:

Canonical link: https://commits.webkit.org/264830@main
